### PR TITLE
Add helm chart repository

### LINKED ISSUE #2311: Helm chart repository
**What would you like to be added**:
A Helm chart repository, whether hosted on GitHub or something like artifacthub.io

**Why is this needed**:
To easily consume the Helm chart via kustomize or other configuration management tools

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by IrvingMg**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
Add Helm chart repository.

#### Which issue(s) this PR fixes:

Fixes #2311 

#### Does this PR introduce a user-facing change?

```release-note
Publish Helm charts in registry.k8s.io/kueue/charts repository
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -37,6 +37,9 @@ IMAGE_REGISTRY ?= $(STAGING_IMAGE_REGISTRY)
 IMAGE_NAME := kueue
 IMAGE_REPO ?= $(IMAGE_REGISTRY)/$(IMAGE_NAME)
 IMAGE_TAG ?= $(IMAGE_REPO):$(GIT_TAG)
+HELM_CHART_REGISTRY ?= $(IMAGE_REGISTRY)
+HELM_CHART_NAME := kueue-$(GIT_TAG).tgz
+HELM_CHART_REPO := charts
```
Comment (trasc) [inline Makefile:42]:
```suggestion
HELM_CHART_REPO := $(IMAGE_REGISTRY)/charts
```

File: Makefile [hunk 2]
```diff
@@ -209,6 +212,14 @@ image-build:
...
 image-push: image-build
 
+.PHONY: helm-chart-build
+helm-chart-build: helm
+	$(HELM) package --version $(GIT_TAG) --app-version $(GIT_TAG) charts/kueue
+
+.PHONY: helm-chart-push
+helm-chart-push: helm-chart-build
+	$(HELM) push ${HELM_CHART_NAME} oci://${HELM_CHART_REGISTRY}/${HELM_CHART_REPO}
+
```
Comment (trasc) [inline Makefile:222]:
```suggestion
.PHONY: helm-chart-push
helm-chart-push: helm
	$(HELM) package --version $(GIT_TAG) --app-version $(GIT_TAG) charts/kueue  -d bin/
	$(HELM) push artifacts/kueue-$(GIT_TAG).tgz oci://${HELM_CHART_REPO}

```

**Review by trasc — COMMENTED** [review]
nits

File: hack/push-chart.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+readonly semver_regex='^v?([0-9]+)(\.[0-9]+)?(\.[0-9]+)?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?$'
+chart_version=${GIT_TAG}
+
+if [[ ${EXTRA_TAG} =~ ${semver_regex} ]]
+then
+	chart_version=${EXTRA_TAG}
+	# Update the image tag and policy
+	${YQ}  e  ".controllerManager.manager.image.repository = \"${IMAGE_REPO}\" | .controllerManager.manager.image.tag = \"${chart_version}\" | .controllerManager.manager.image.pullPolicy = \"IfNotPresent\"" -i charts/kueue/values.yaml 
+fi
```
Comment (trasc) [inline hack/push-chart.sh:35]: this should be reverted after the build, otherwise the source tree will remain dirty

File: hack/push-chart.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+YQ=${YQ:-./bin/yq}
+
+readonly semver_regex='^v?([0-9]+)(\.[0-9]+)?(\.[0-9]+)?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?$'
+chart_version=${GIT_TAG}
+
+if [[ ${EXTRA_TAG} =~ ${semver_regex} ]]
+then
+	chart_version=${EXTRA_TAG}
+	# Update the image tag and policy
+	${YQ}  e  ".controllerManager.manager.image.repository = \"${IMAGE_REPO}\" | .controllerManager.manager.image.tag = \"${chart_version}\" | .controllerManager.manager.image.pullPolicy = \"IfNotPresent\"" -i charts/kueue/values.yaml
```
Comment (trasc) [inline hack/push-chart.sh:34]: IMAGE_REPO should be `registry.k8s.io/kueue` in this case

File: hack/push-chart.sh [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+EXTRA_TAG=${EXTRA_TAG:-$(git branch --show-current)} 
+GIT_TAG=${GIT_TAG:-$(git describe --tags --dirty --always)}
+HELM_CHART_REPO=${HELM_CHART_REPO:-gcr.io/k8s-staging-kueue/charts}
+IMAGE_REPO=${IMAGE_REPO:-gcr.io/k8s-staging-kueue/kueue}
+
+HELM=${HELM:-./bin/helm}
+YQ=${YQ:-./bin/yq}
+
+readonly semver_regex='^v?([0-9]+)(\.[0-9]+)?(\.[0-9]+)?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?$'
```
Comment (trasc) [inline hack/push-chart.sh:27]:
```suggestion
readonly semver_regex='^v([0-9]+)(\.[0-9]+){1,2}$'
```

**Review by trasc — COMMENTED** [review]
After this is merged and we have "production confirmation" that it works as we expect we showed follow-up on:

- check if we are also able to promote the charts to registry.k8s.io
- drop the old build method (release artifacts) 
- update the install related documentation

/lgtm

**Comment by mruoss** [comment]
> After this is merged and we have "production confirmation" that it works as we expect we showed follow-up on:
> 
> * check if we are also able to promote the charts to registry.k8s.io
> * drop the old build method (release artifacts)
> * update the install related documentation
> 
> /lgtm

 A package publication on https://artifacthub.io/ whould be nice, too.

**Comment by alculquicondor** [comment]
/release-note-edit

```release-note
Publish Helm charts in registry.k8s.io/kueue/charts repository
```

Assuming that we can promote the image there. If it doesn't work, we can revert this PR.

**Comment by alculquicondor** [comment]
It looks like the objects are properly listed in gcr.io/k8s-staging-kueue/charts/kueue

Should it be gcr.io/k8s-staging-kueue/kueue/charts instead?

**Comment by alculquicondor** [comment]
Also, can you test those objects to see if they work properly?

**Comment by alculquicondor** [comment]
I left a question in the k8s-infra slack channel https://kubernetes.slack.com/archives/CCK68P2Q2/p1718819420490889

**Comment by BenTheElder** [comment]
Nobody should be consuming from gcr.io/k8s-staging-kueue directly anyhow (in fact we will have to move to artifact registry later this year) so the path doesn't matter all that much there, when promoted to registry.k8s.io the path can be different.

**Comment by alculquicondor** [comment]
From slack thread:

> Please start publishing staging images for kueue to us-central1-docker.pkg.dev/k8s-staging-images/kueue

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
Experimental support for helm charts in the gcr.io/k8s-staging-kueue/charts/kueue repository
```

Because #2476 didn't merge and I don't know if it's safe to do so.

**Pull request merged**